<a href="https://colab.research.google.com/github/unatardedemartes-a11y/ClassFiles/blob/main/Actividad_Calificada_07_Datos_Duplicados_275334.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** |Livia Daniela Padilla Barragan |
| **Matrícula** |275334|

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [34]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [35]:
# Escribe tu código aquí
dup_first= hoteles.duplicated(keep='first').sum() #marca el siguiente registro que esta duplicado como false
dup_last= hoteles.duplicated(keep='last').sum()   #marca el ultimo registro que esta duplicado como false
dup_false= hoteles.duplicated(keep=False).sum()   #marca todos los registros que tienen una duplicidad
print("Duplicados first:",dup_first)
print("Duplicados last:",dup_last)
print("Duplicados false:",dup_false)

Duplicados first: 31994
Duplicados last: 31994
Duplicados false: 40165


### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [36]:
# Escribe tu código aquí
total_filas = len(hoteles)
porcentaje_sobrantes = (dup_first / total_filas) * 100  #porcentaje de filas sobrantes respecto al total de filas
print("Porcentaje de filas sobrantes:", round(porcentaje_sobrantes,2),"%")

grupos_repetidos = (dup_false - dup_first) // 2  #numero de grupos repetidos (combinaciones de valores que aparecen más de una vez
print("Número de grupos repetidos:", grupos_repetidos)

Porcentaje de filas sobrantes: 26.8 %
Número de grupos repetidos: 4085


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [37]:
# Escribe tu código aquí

hoteles_dup = hoteles.copy() #hacer copia del dataset

hoteles_dup['es_duplicado'] = hoteles_dup.duplicated(keep='first') #agregar columna para identificar duplicados

total_por_hotel = hoteles_dup['hotel'].value_counts() #total de reservcas por hotel

sobrantes_por_hotel = hoteles_dup.loc[hoteles_dup['es_duplicado'], 'hotel'].value_counts() ##Filas sobrantes (duplicadas) por hotel
porcentaje_por_hotel = sobrantes_por_hotel / total_por_hotel * 100
print("Sobrantes por hotel")
print(sobrantes_por_hotel)
print()
print("Porcentaje por hotel")
print(round(porcentaje_por_hotel,2))


Sobrantes por hotel
hotel
City Hotel      25902
Resort Hotel     6092
Name: count, dtype: int64

Porcentaje por hotel
hotel
City Hotel      32.65
Resort Hotel    15.21
Name: count, dtype: float64


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**

El problema es grave y no afecta por igual a ambos hoteles: City Hotel duplica un 32.65 % de sus reservas, más del doble que Resort Hotel (15.21 %)

---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [38]:
# Escribe tu código aquí
columnas_ver = [
    'hotel', 'lead_time',
    'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month',
    'adults', 'country', 'market_segment', 'deposit_type',
    'adr', 'reservation_status'
]

duplicados = hoteles[hoteles.duplicated(keep=False)]
duplicados_ordenados = duplicados.sort_values(by=list(hoteles.columns)) #ordenamos para que las copias queden juntas
                                                                        #para poder ver los grupos claramente

print("Primeras 12 filas involucradas en duplicados (keep=False)")
print(duplicados_ordenados[columnas_ver].head(12).to_string())

Primeras 12 filas involucradas en duplicados (keep=False)
            hotel  lead_time  arrival_date_year arrival_date_month  arrival_date_day_of_month  adults country market_segment deposit_type   adr reservation_status
40772  City Hotel          0               2015             August                          7       2     PRT         Direct   No Deposit  75.0          Check-Out
40802  City Hotel          0               2015             August                          7       2     PRT         Direct   No Deposit  75.0          Check-Out
40821  City Hotel          0               2015             August                          8       2     PRT      Online TA   No Deposit  89.0          Check-Out
40838  City Hotel          0               2015             August                          8       2     PRT      Online TA   No Deposit  89.0          Check-Out
76792  City Hotel          0               2015             August                         10       2     FRA      Online TA   

### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [39]:
# Escribe tu código aquí
conteo_combinaciones = hoteles.value_counts(dropna=False) #contamos las combinaciones y usamos el dropna
                                                          #porque 'agent' y 'company' tienen NaN

grupos_por_tamano = conteo_combinaciones.value_counts().sort_index() #ordena los grupos por tamano
print("Grupos por tamaño (nº de copias)")
print(grupos_por_tamano.head(10))

tamano_max = conteo_combinaciones.max() #grupo de mayor tamano
combinacion_max = conteo_combinaciones.idxmax()
print()
print("Combinación de valores:")
print(pd.Series(combinacion_max, index=hoteles.columns))

Grupos por tamaño (nº de copias)
count
1     79225
2      5034
3       969
4       497
5       275
6       171
7       115
8       103
9        93
10      123
Name: count, dtype: int64

Combinación de valores:
hotel                             City Hotel
is_canceled                                1
lead_time                                277
arrival_date_year                       2016
arrival_date_month                  November
arrival_date_week_number                  46
arrival_date_day_of_month                  7
stays_in_weekend_nights                    1
stays_in_week_nights                       2
adults                                     2
children                                 0.0
babies                                     0
meal                                      BB
country                                  PRT
market_segment                        Groups
distribution_channel                   TA/TO
is_repeated_guest                          0
previous_cancellations   

### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [40]:
# Escribe tu código aquí

completo   = hoteles
duplicados = hoteles[hoteles.duplicated(keep=False)]

def distribucion(df, col):
    return df[col].value_counts(normalize=True).sort_index()

# --- market_segment ---
ms = pd.DataFrame({
    'completo_%'  : (distribucion(completo,   'market_segment') * 100).round(2),
    'duplicados_%': (distribucion(duplicados, 'market_segment') * 100).round(2),
})
ms['dif_pp'] = (ms['duplicados_%'] - ms['completo_%']).round(2)
ms = ms.sort_values('completo_%', ascending=False)
print("market_segment")
print(ms.to_string())

# --- deposit_type ---
dt = pd.DataFrame({
    'completo_%'  : (distribucion(completo,   'deposit_type') * 100).round(2),
    'duplicados_%': (distribucion(duplicados, 'deposit_type') * 100).round(2),
})
dt['dif_pp'] = (dt['duplicados_%'] - dt['completo_%']).round(2)
dt = dt.sort_values('completo_%', ascending=False)
print("deposit_typ")
print(dt.to_string())

# --- is_canceled ---
print("Tasa de cancelación")
print("Completo   :", round(completo['is_canceled'].mean()*100,2), "%")
print("Duplicados :", round(duplicados['is_canceled'].mean()*100,2),"%")

market_segment
                completo_%  duplicados_%  dif_pp
market_segment                                  
Online TA            47.30         20.59  -26.71
Offline TA/TO        20.29         30.40   10.11
Groups               16.59         41.61   25.02
Direct               10.56          3.50   -7.06
Corporate             4.44          3.67   -0.77
Complementary         0.62          0.19   -0.43
Aviation              0.20          0.05   -0.15
Undefined             0.00           NaN     NaN
deposit_typ
              completo_%  duplicados_%  dif_pp
deposit_type                                  
No Deposit         87.65         64.17  -23.48
Non Refund         12.22         35.64   23.42
Refundable          0.14          0.19    0.05
Tasa de cancelación
Completo   : 37.04 %
Duplicados : 58.39 %


### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

1. Es mas probable que sean erroes debido a la coincidencia en todas las columnas, pero no es concluyente al 100 % por la ausencia de un ID de reserva.
2. Si dos filas tienen agent = NaN y company = NaN y el resto de columnas idénticas, duplicated() las marca como duplicadas. Esto es lo deseable aquí porque queremos detectar cualquier fila repetida, tenga o no faltantes

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [41]:
# Escribe tu código aquí
filas_antes = len(hoteles)

hoteles_limpio = hoteles.drop_duplicates(keep='first', ignore_index=True) #Eliminamos duplicados exactos conservando la primera aparición
                                                                          #y donde ignore_index=True para que renumere el índice desde 0
filas_despues = len(hoteles_limpio)

filas_eliminadas = filas_antes - filas_despues
print("Filas antes:", filas_antes)
print("Filas después:", filas_despues)
print("Filas eliminadas:", filas_eliminadas)

Filas antes: 119390
Filas después: 87396
Filas eliminadas: 31994


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [42]:
# Escribe tu código aquí
#dataset original
reservas_orig   = hoteles.groupby('hotel').size()
medias_orig     = hoteles.groupby('hotel')[['is_canceled', 'adr']].mean()

metricas_orig = pd.DataFrame({
    'reservas'          : reservas_orig,
    'tasa_cancelacion'  : medias_orig['is_canceled'],
    'adr_promedio'      : medias_orig['adr'],
})

#dataset sin duplicados
reservas_limpio = hoteles_limpio.groupby('hotel').size()
medias_limpio   = hoteles_limpio.groupby('hotel')[['is_canceled', 'adr']].mean()

metricas_limpio = pd.DataFrame({
    'reservas'          : reservas_limpio,
    'tasa_cancelacion'  : medias_limpio['is_canceled'],
    'adr_promedio'      : medias_limpio['adr'],
})

print("MÉTRICAS CON DUPLICADOS (hoteles)")
print(metricas_orig.round(2).to_string())
print()
print("MÉTRICAS SIN DUPLICADOS (hoteles_limpio) ")
print(metricas_limpio.round(2).to_string())

MÉTRICAS CON DUPLICADOS (hoteles)
              reservas  tasa_cancelacion  adr_promedio
hotel                                                 
City Hotel       79330              0.42        105.30
Resort Hotel     40060              0.28         94.95

MÉTRICAS SIN DUPLICADOS (hoteles_limpio) 
              reservas  tasa_cancelacion  adr_promedio
hotel                                                 
City Hotel       53428              0.30        110.99
Resort Hotel     33968              0.23         99.03


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [43]:
# Escribe tu código aquí

#Resumen del dataset ORIGINAL
metricas_orig = (
    hoteles
    .groupby('hotel')
    .agg(
        reservas         = ('is_canceled', 'size'),
        tasa_cancelacion = ('is_canceled', 'mean'),
        adr_promedio     = ('adr', 'mean'),
    )
)
metricas_orig['version'] = 'original'

#Resumen del dataset LIMPIO
metricas_limpio = (
    hoteles_limpio
    .groupby('hotel')
    .agg(
        reservas         = ('is_canceled', 'size'),
        tasa_cancelacion = ('is_canceled', 'mean'),
        adr_promedio     = ('adr', 'mean'),
    )
)
metricas_limpio['version'] = 'sin duplicados'

#Unión con pd.concat()
comparacion = pd.concat([metricas_orig, metricas_limpio])

#Índice renumerado y ordenado por hotel
comparacion = (
    comparacion
    .reset_index()                              # 'hotel' pasa a ser columna
    .sort_values(by=['hotel', 'version'])       # agrupa las dos versiones de cada hotel
    .reset_index(drop=True)                     # renumera el índice desde 0
    .round(4)
)

print("TABLA COMPARATIVA")
print(comparacion.to_string())

TABLA COMPARATIVA
          hotel  reservas  tasa_cancelacion  adr_promedio         version
0    City Hotel     79330            0.4173      105.3045        original
1    City Hotel     53428            0.3004      110.9859  sin duplicados
2  Resort Hotel     40060            0.2776       94.9529        original
3  Resort Hotel     33968            0.2348       99.0253  sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

1. La métrica que más cambia es el número de reservas, y el cambio es mucho mayor en City Hotel (~32.65 % eliminado frente a ~15.21 % en Resort Hotel).
2. Métricas infladas, validación cruzada inválida y comparaciones sesgadas entre modelos

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [44]:
# Escribe tu código aquí
#duplicados restantes en limpio
dup_restantes = hoteles_limpio.duplicated().sum()
print("Duplicados exactos restantes:",dup_restantes)

#indices repetidos
indices_repetidos = hoteles_limpio.index.duplicated().sum()
print("Índices repetidos:",indices_repetidos)

#El índice va de 0 a len-1 de forma consecutiva?
n = len(hoteles_limpio)
indice_esperado = pd.RangeIndex(start=0, stop=n, step=1)

indice_correcto = hoteles_limpio.index.equals(indice_esperado)
print("Longitud del dataset        :", n)
print("Primer índice               :", hoteles_limpio.index[0])
print("Último índice               :", hoteles_limpio.index[-1])


Duplicados exactos restantes: 0
Índices repetidos: 0
Longitud del dataset        : 87396
Primer índice               : 0
Último índice               : 87395


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [45]:
# Escribe tu código aquí
#1. Todas las columnas menos reservation_status_date
columnas_clave = [c for c in hoteles_limpio.columns if c != 'reservation_status_date']

#filas que coinciden en todas las columnas clave (keep=False)
mascara_parcial = hoteles_limpio.duplicated(subset=columnas_clave, keep=False)

parciales = hoteles_limpio[mascara_parcial].copy()

sobrantes_parciales = hoteles_limpio.duplicated(subset=columnas_clave, keep='first').sum()
total_involucradas  = mascara_parcial.sum()

print("Filas involucradas en duplicados parciales (keep=False) :", total_involucradas)
print("Filas sobrantes (keep='first')                          :", sobrantes_parciales)

#2. Ordenar filas por x columnas

columnas_ver = [
    'hotel', 'lead_time',
    'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month',
    'adr', 'reservation_status', 'reservation_status_date'
]

# Ordenamos por las columnas clave + la fecha, para que las copias queden juntas
parciales_ordenados = parciales.sort_values(
    by=columnas_clave + ['reservation_status_date']
)
print()
print("Primeras 15 filas de duplicados parciales")
print(parciales_ordenados[columnas_ver].head(15).to_string())

#3. Distribución de 'reservation_status'
dist_status = parciales['reservation_status'].value_counts()
dist_status_pct = parciales['reservation_status'].value_counts(normalize=True) * 100
print()
print("Distribución de reservation_status (filas duplicadas parciales)")
print(pd.DataFrame({
    'conteo'    : dist_status,
    'porcentaje': dist_status_pct.round(2)
}).to_string())

Filas involucradas en duplicados parciales (keep=False) : 471
Filas sobrantes (keep='first')                          : 245

Primeras 15 filas de duplicados parciales
            hotel  lead_time  arrival_date_year arrival_date_month  arrival_date_day_of_month     adr reservation_status reservation_status_date
35574  City Hotel          9               2015          September                         10   95.00           Canceled              2015-09-01
35572  City Hotel          9               2015          September                         10   95.00           Canceled              2015-09-03
51268  City Hotel         12               2017               July                         15  139.00           Canceled              2017-07-07
51267  City Hotel         12               2017               July                         15  139.00           Canceled              2017-07-10
54170  City Hotel         14               2015            October                         15  100.00       

**Respuesta 4.2:**

EL reservation_status_date podria tratarse de un campo para registrar fechas efectivas de acuerdo al cambio de reservation status, lo que explicaria porque tenemos registros casi identicos, por lo que no las eliminaria

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

Criterio: se compararon todas las columnas (duplicated() sin subset), porque no hay ID de reserva. Los NaN de agent y company se tratan como iguales entre sí.

Conservación: keep='first'. Da igual first o last porque las copias son idénticas.

Resultado: City Hotel pierde 32.65 %, Resort Hotel 15.21 %. Verificado: sin duplicados exactos, sin índices repetidos

Supuestos: coincidir en todas las columnas = mismo registro. La mayoría son errores técnicos (doble carga), no reservas distintas, por el volumen y la concentración en City Hotel.

Limitaciones: sin ID no se puede identificar si es un duplicado técnico de re-reserva real. 'Company' tiene muchos faltantes y discrimina poco.

---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

1. Si cada fila representa un evento real (no una reserva única), por ejemplo dos clientes distintos que hacen la misma reserva el mismo día, mismo hotel, mismos adultos y mismo ADR (tarifa fija). En ese caso son reservas legítimas y borrarlas subestimaría la demanda.

2. Columnas adicionales que pediría:

* ID de reserva para saber si es el mismo registro o dos distintos.

* Timestamp de creación/actualización para distinguir doble carga de eventos separados en el tiempo.

* ID de cliente para verificar si es la misma persona

3. duplicated().sum() == 0 no garantiza nada porque solo detecta duplicados exactos en las columnas comparadas

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.